gr-network/lib/tcp_source_impl.cc

```diff
@@ -0,0 +1,329 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2025 Skandalis Georgios
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+#include "gnuradio/block.h"
+#include "gnuradio/network/tcp_source.h"
+#include "gnuradio/sptr_magic.h"
+#include "gnuradio/types.h"
+#include <cerrno>
+#include <cstdlib>
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+
+#if defined(_WIN32) || defined(_WIN64)
+
+#define WINDOWS_SOCKETS
+
+#pragma commend(lib, "Ws2_32.lib")
+
+#include <Winsock2.h>
+#include <ws2tcpip.h>
+
+#else
+
+#include <arpa/inet.h>
+#include <asm-generic/socket.h>
+#include <netdb.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
+
+#endif
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "tcp_source_impl.h"
+
+namespace gr {
+namespace network {
+
+typedef struct addrinfo addrinfo_t;
+
+tcp_source::sptr tcp_source::make(size_t itemsize,
+                                  size_t veclen,
+                                  const std::string& host,
+                                  const std::string& port,
+                                  int source_mode)
+{
+    return gnuradio::make_block_sptr<tcp_source_impl>(
+        itemsize, veclen, host, port, source_mode);
+}
+
+tcp_source_impl::tcp_source_impl(size_t item_size,
+                                 size_t vec_len,
+                                 const std::string& host,
+                                 const std::string& port,
+                                 int source_mode)
+    : d_connection_mode(source_mode), d_block_size(item_size * vec_len), d_veclen(vec_len)
+{
+#ifdef WINDOWS_SOCKETS
+
+    WSADATA wsaData;
+
+    int wsa_res = WSAStartup(MAKEWORD(2, 2), &wsaData);
+
+    if (wsa_res != 0) {
+        d_logger->error("WSAStartup failed with error code {}.", res);
+        throw std::runtime_error("Error while initializing Windows sockets.");
+    }
+
+    if (LOBYTE(wsaData.wVersion) != 2 || HIBYTE(wsaData.wVersion) != 2) {
+        d_logger->error("Winsock version 2.2 not available.");
+        WSACleanup();
+        throw std::runtime_error("Error while initializing Windows sockets.");
+    }
+
+#endif
+
+    struct addrinfo hints;
+    // very ugly but its either that or make an entire class for a single pointer that
+    // gets used a single time
+    std::unique_ptr<addrinfo_t, void (*)(addrinfo_t*)> host_info(new addrinfo_t,
```

Comment:
[spec](https://en.cppreference.com/w/cpp/memory/unique_ptr/reset.html) says deleter is only called for non-null, so this could be `nullptr` instead of a dummy allocation.

In fact, it may be safer, because `freeaddrinfo()` could quite possibly assume not just a valid pointer, but a valid object.

Reply:
Could you explain this in simpler terms? I'm not entirely familiar with a lot of C++ specifics so this is hard for me to understand

Reply:
`new addrinfo_t` just creates a struct, it'll exist in memory but will have garbage data. E.g. `host_info->ai_next` will be a dangling pointer.

`freeaddrinfo()` may (in fact probably will) try to follow that pointer.

Reply:
So, if I understand it correctly instead of `new addrinfo_t` I should use `nullptr` in the constructor, no?